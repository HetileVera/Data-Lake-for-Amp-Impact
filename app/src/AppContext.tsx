import React, { createContext, useContext } from "react";
import isEmpty from "lodash/isEmpty";
import throttle from "lodash/throttle";

import { AppStore, IAppStore } from "./models/AppStore";
import { IInstallation } from "./models/Installation";
import { isLocalDev, websiteUrlOrigin, assetWebsiteUrl, secretKey } from "./helpers/settings";
import { onSnapshot } from "mobx-state-tree";

// Create an instance of the root store (a.k.a appStore)
const appStore = AppStore.create({});

// Store state to local storage
const storeLocally = throttle(() => appStore.saveLocal(), 5000);

// Store state in S3 when applicable
const storeRemotely = throttle(() => appStore.saveRemote(), 30000);

// The initialization function for the root store
export function initializeStore() {
  if (!isEmpty(secretKey)) console.warn("Access key is bundled!");

  if (isLocalDev) {
    console.log(`isLocalDev '${isLocalDev}'`);
    console.log(`websiteUrlOrigin '${websiteUrlOrigin}'`);
    console.log(`assetWebsiteUrl '${assetWebsiteUrl}'`);

    if (window.location.host !== "localhost:3000") {
      console.warn("You should run the app on localhost!");

      throw new Error("App should run on localhost for development mode!");
    }
  }

  appStore.loadLocal();
  onSnapshot(appStore, (snapshot) => {
    storeLocally();
    storeRemotely();
  });
}

// Create the store context
export const StoreContext = createContext<IAppStore | undefined>(undefined);

// Create the store provider component
export const StoreProvider: React.FunctionComponent = ({ children }) => {
  return <StoreContext.Provider value={appStore}>{children}</StoreContext.Provider>;
};

// useStore hook
export function useStore(): IAppStore {
  const store = useContext(StoreContext);

  if (!store) {
    throw new Error("useStore() must be used within a StoreProvider");
  }

  return store;
}

// useInstallation hook
export function useInstallation(): IInstallation {
  const store = useStore();
  const installation = store.installation;

  if (!installation) {
    throw new Error("An installation instance was not created yet");
  }

  return installation;
}
