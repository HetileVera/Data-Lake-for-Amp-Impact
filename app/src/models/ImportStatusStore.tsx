import includes from "lodash/includes";
import split from "lodash/split";
import last from "lodash/last";
import chunkIt from "lodash/chunk";
import first from "lodash/first";
import isEmpty from "lodash/isEmpty";
import sortBy from "lodash/sortBy";
import { ExecutionStatus, HistoryEvent } from "@aws-sdk/client-sfn";
import { types, Instance, getParent } from "mobx-state-tree";
import { values } from "mobx";
import { useEffect } from "react";

import { useInstallation } from "AppContext";
import { s3GetJson } from "api/s3-get-json";
import { describeSFN } from "api/describe-sfn";
import { getExecutionFailureInfo, getSFNHistory, isExecutionFailureEvent } from "api/get-sfn-history";
import { s3ListObjects, S3Object } from "api/list-s3-objects";
import { getLatestSFN } from "api/get-lastest-sfn";
import { delay, niceByte, niceNumber } from "helpers/utils";

import { BaseStore, isStoreError, isStoreLoading, isStoreNew, isStoreReady, isStoreReLoading } from "./BaseStore";
import { ICredentials } from "./helpers/Credentials";
import { IInstallation } from "./Installation";

const isSuccessStatus = (status: ExecutionStatus) => {
  // For all statuses, see https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html

  // We consider any of these statuses as a success
  const successStatuses = ["SUCCEEDED"];
  return includes(successStatuses, status);
};

const isFailureStatus = (status: ExecutionStatus) => {
  // For all statuses, see https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html

  // We consider any of these statuses as a failure
  const failureStatuses = ["FAILED", "ABORTED", "TIMED_OUT"];
  return includes(failureStatuses, status);
};

export enum ObjectImportStatusCode {
  Not_Started = "NOT_STARTED",
  Preparing = "PREPARING",
  In_Progress = "IN_PROGRESS",
  Importing = "IMPORTING",
  Successful = "SUCCESSFUL",
}

interface ObjectStatusInput {
  name: string;
  label?: string;
  status?: ObjectImportStatusCode;
  rowCount?: number;
  columnCount?: number;
  importSeconds?: number;
  bytesWritten?: number;
}

interface RemoteObject {
  status: ObjectImportStatusCode;
  metadata?: {
    rowCount: number;
    columnCount: number;
    importSeconds: number;
    bytesWritten: number;
  };
}

/**
 * Represents a Salesforce object import status
 */
const ObjectImportStatus = types
  .model("ObjectImportStatus", {
    name: "",
    label: "",
    status: types.optional(
      types.enumeration<ObjectImportStatusCode>("ObjectImportStatusCode", Object.values(ObjectImportStatusCode)),
      ObjectImportStatusCode.Not_Started
    ),
    rowCount: 0,
    columnCount: 0,
    importSeconds: 0,
    bytesWritten: 0,
    stale: false,
  })

  .actions((self) => ({
    setObjectStatus({ name, label, status, rowCount, columnCount, importSeconds, bytesWritten }: ObjectStatusInput) {
      self.name = name;
      if (label) self.label = label;
      if (status) self.status = status;
      if (rowCount) self.rowCount = rowCount;
      if (columnCount) self.columnCount = columnCount;
      if (importSeconds) self.importSeconds = importSeconds;
      if (bytesWritten) self.bytesWritten = bytesWritten;
    },

    setStale(flag: boolean) {
      self.stale = flag;
      if (flag) self.status = ObjectImportStatusCode.Not_Started;
    },
  }))

  .views((self) => ({
    get isNotStarted(): boolean {
      return self.status === ObjectImportStatusCode.Not_Started;
    },

    get isPreparing(): boolean {
      return self.status === ObjectImportStatusCode.Preparing;
    },

    get isInProgress(): boolean {
      return self.status === ObjectImportStatusCode.In_Progress;
    },

    get isImporting(): boolean {
      return self.status === ObjectImportStatusCode.Importing;
    },

    get isSuccessful(): boolean {
      return self.status === ObjectImportStatusCode.Successful;
    },

    get warning(): boolean {
      return self.rowCount < 0;
    },
  }))

  .views((self) => ({
    get statsAvailable(): boolean {
      // Tells us if any of the stats such as (rowCount, et) are available.
      return self.isSuccessful;
    },
  }))

  .views((self) => ({
    get rowCountDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceNumber(self.rowCount);
    },

    get columnCountDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceNumber(self.columnCount);
    },

    get importSecondsDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return `${Math.round(self.importSeconds)} sec`;
    },

    get bytesWrittenDisplay(): string {
      if (!self.statsAvailable) return "N/A";

      return niceByte(self.bytesWritten);
    },
  }));

export interface IObjectImportStatus extends Instance<typeof ObjectImportStatus> {}

export enum LoadingStage {
  GettingSchemas = "GETTING_SCHEMAS",
  MonitoringProgress = "MONITORING_PROGRESS",
}

/**
 * Represents the model for the status of importing salesforce objects. This includes each object import status.
 */
export const ImportStatusStore = BaseStore.named("ImportStatusStore")
  .props({
    objects: types.map(ObjectImportStatus),
    startTime: 0, // The start time of the execution (epoch milliseconds)
    loadingStage: types.maybe(types.enumeration<LoadingStage>("LoadingStage", Object.values(LoadingStage))),
    _errorDetail: "", // Expected to be in markdown https://commonmark.org/help/
  })

  .views((self) => ({
    get empty() {
      return self.objects.size === 0;
    },

    get credentials(): ICredentials {
      const parent: IInstallation = getParent(self);
      return parent.credentials;
    },

    get region(): string {
      const parent: IInstallation = getParent(self);
      return parent.region;
    },

    get importWorkflowArn(): string {
      const parent: IInstallation = getParent(self);
      return parent.importWorkflowArn;
    },

    get metadataBucket(): string {
      const parent: IInstallation = getParent(self);
      return parent.metadataBucket;
    },

    get stateMachineArn(): string {
      const parent: IInstallation = getParent(self);
      return parent.importWorkflowArn;
    },

    get isGettingSchemas() {
      return self.loadingStage === LoadingStage.GettingSchemas && (isStoreLoading(self) || isStoreReLoading(self));
    },

    get isMonitoringProgress() {
      return self.loadingStage === LoadingStage.MonitoringProgress && (isStoreLoading(self) || isStoreReLoading(self));
    },

    get errorDetail() {
      return self._errorDetail;
    },

    get listObjects(): IObjectImportStatus[] {
      const result: readonly unknown[] = values(self.objects);
      return sortBy(result as IObjectImportStatus[], ["label", "name"]);
    },

    get objectsCount(): number {
      return self.objects.size;
    },

    get processPercentage(): number {
      const size = self.objects.size;
      let count = 0;

      self.objects.forEach((object) => {
        switch (object.status) {
          case ObjectImportStatusCode.Preparing:
            count += 0.25;
            break;
          case ObjectImportStatusCode.In_Progress:
            count += 0.5;
            break;
          case ObjectImportStatusCode.Importing:
            count += 0.75;
            break;
          case ObjectImportStatusCode.Successful:
            count += 1;
        }
      });

      if (size === 0) return 0; // Avoid division by zero
      return (count / size) * 100;
    },
  }))

  .views((self) => ({
    get isSuccess(): boolean {
      return isStoreReady(self) && !isStoreError(self) && isEmpty(self.errorDetail);
    },

    get totalRecords(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).rowCount, 0), 0);
    },

    get totalFields(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).columnCount, 0), 0);
    },

    get totalBytes(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).bytesWritten, 0), 0);
    },

    get totalTime(): number {
      return self.listObjects.reduce<number>((result, item) => result + Math.max((item as IObjectImportStatus).importSeconds, 0), 0);
    },
  }))

  .views((self) => ({
    get totalRecordsDisplay(): string {
      return niceNumber(self.totalRecords);
    },

    get totalFieldsDisplay(): string {
      return niceNumber(self.totalFields);
    },

    get totalBytesDisplay(): string {
      return niceByte(self.totalBytes);
    },

    get totalTimeDisplay(): string {
      return `${Math.round(self.totalTime)} sec`;
    },
  }))

  .actions((self) => ({
    setObject(rawObject: ObjectStatusInput) {
      if (self.objects.has(rawObject.name)) {
        self.objects.get(rawObject.name)?.setObjectStatus(rawObject);
      } else {
        self.objects.set(rawObject.name, rawObject);
      }

      return self.objects.get(rawObject.name);
    },

    setLoadingStage(stage: LoadingStage) {
      self.loadingStage = stage;
    },

    clearLoadingStage() {
      self.loadingStage = undefined;
    },

    setStartTime(time: number) {
      self.startTime = time;
    },
  }))

  .actions((self) => ({
    buildStateMachineUrl(): string {
      const region = self.region;
      const stateMachineArn = self.importWorkflowArn;

      return `https://${region}.console.aws.amazon.com/states/home?region=${region}#/statemachines/view/${encodeURIComponent(stateMachineArn)}`;
    },

    buildExecutionUrl(executionArn: string): string {
      const region = self.region;

      return `https://${region}.console.aws.amazon.com/states/home?region=${region}#/executions/details/${encodeURIComponent(executionArn)}`;
    },
  }))

  .actions((self) => ({
    async describeWorkflow(executionArn: string): Promise<ExecutionStatus> {
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const workflowArn = self.importWorkflowArn;

      try {
        const info = await describeSFN({ accessKey, secretKey, region, executionArn });
        return info.status as ExecutionStatus;
      } catch (error) {
        self.runInAction(() => {
          // This is in markdown
          self._errorDetail = `We attempted to check the status of the execution of the state machine by calling the Step Functions DescribeExecution API.
          However, we got the following error:
### ${error}

### Information
* State machine execution ARN: ${executionArn}
* State machine ARN: ${workflowArn}

### Links
* [AWS Step Functions execution page](${self.buildExecutionUrl(executionArn)})
* [AWS Step Functions state machine page](${self.buildStateMachineUrl()})
* [Step Functions DescribeExecution API documentation](https://docs.aws.amazon.com/step-functions/latest/apireference/API_DescribeExecution.html)
          `;
        });
        throw error;
      }
    },

    async handleSFNFailure(executionArn: string, status: ExecutionStatus) {
      // In this method we display an informative error detail that provides enough context around the failure.
      // To do this, we will list all the execution history events and look for the overall execution event that had an error.
      // Since getting the history events is not critical, we will capture any errors from calling GetExecutionHistory
      // and not propagate them or have them affecting the overall status of the operation.

      let executionEvent: HistoryEvent;
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const workflowArn = self.importWorkflowArn;

      try {
        const events = await getSFNHistory({ accessKey, secretKey, region, executionArn });
        // We find the execution event that failed
        for (const event of events) {
          if (isExecutionFailureEvent(event)) {
            // We found the event
            executionEvent = event;
            break;
          }
        }
      } catch (error) {
        // Since getting the stack events is not critical, we will capture any errors from calling the describe stack events
        // call and not propagate them or have them affect the overall status of the operation.
        console.log(error);
      }

      self.runInAction(() => {
        // This is in markdown
        self._errorDetail = `The step function workflow failed with the following status:
### ${status}`;

        if (executionEvent) {
          const info = getExecutionFailureInfo(executionEvent);
          self._errorDetail = `${self._errorDetail}
### Execution Failure
* Cause: ${info.cause}
         `;
        }

        self._errorDetail = `${self._errorDetail}
### Information
* State machine execution ARN: ${executionArn}
* State machine ARN: ${workflowArn}
        
### Links
* [AWS Step Functions execution page](${self.buildExecutionUrl(executionArn)})
* [AWS Step Functions state machine page](${self.buildStateMachineUrl()})        
          `;
      });
      throw new Error(status);
    },

    async loadSchemas() {
      // We need to detect all the objects that we are expecting to import, this is done by looking at the schema files in S3
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const bucketName = self.metadataBucket;
      const worker = async (s3Object: S3Object) => {
        const json = (await s3GetJson({ accessKey, secretKey, region, bucketName, objectKey: s3Object.key })) as { name: string; label: string };
        self.setObject({ name: json.name, label: json.label ?? json.name });
      };

      const objects = await s3ListObjects({ accessKey, secretKey, region, bucketName, prefix: `schemas/` });
      const chunks = chunkIt(objects, 5); // chunks them 5 at a time
      while (!isEmpty(chunks)) {
        const chunk = chunks.shift() ?? [];
        // Create a list of worker promises and wait on all of them
        await Promise.all(chunk.map((object) => worker(object)));
      }
    },

    async pullProgress(executionArn: string) {
      // We pull that latest status of any object that has not reached the successful stage
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const runId = last(split(executionArn, ":"));
      const bucketName = self.metadataBucket;
      const worker = async (s3Object: S3Object) => {
        // An example of an S3 object key is
        // sf-metadata-u1puzajv0crm/state/runs/aaaaa-8d51-4dd3-bbf7-a81ee93cdc9a/Account.status.json
        // So, can split on "/" and take the last element which is "Account.status.json" and then
        // split on "." and take the first element which give us "Account"
        const name = first(split(last(split(s3Object.key, "/")), ".")) ?? ""; // Yes, we can regex but this is also okay, see explanation above
        try {
          const existing = self.objects.get(name);
          if (existing?.isSuccessful) return;

          const json = (await s3GetJson({ accessKey, secretKey, region, bucketName, objectKey: s3Object.key })) as RemoteObject;
          const object = self.setObject({
            name,
            status: json.status,
            rowCount: json.metadata?.rowCount,
            columnCount: json.metadata?.columnCount,
            importSeconds: json.metadata?.importSeconds,
            bytesWritten: json.metadata?.bytesWritten,
          });

          object?.setStale(false);
        } catch (error) {
          // NOTE: it is important that we don't propagate this error, we don't want status reporting to halt deployment
          console.error(error);
        }
      };

      const objects = await s3ListObjects({ accessKey, secretKey, region, bucketName, prefix: `state/runs/${runId}/` });
      const chunks = chunkIt(objects, 5); // chunks them 5 at a time
      while (!isEmpty(chunks)) {
        const chunk = chunks.shift() ?? [];
        // Create a list of worker promises and wait on all of them
        await Promise.all(chunk.map((object) => worker(object)));
      }
    },

    async getLatestExecution() {
      const accessKey = self.credentials.accessKey;
      const secretKey = self.credentials.secretKey;
      const region = self.region;
      const stateMachineArn = self.stateMachineArn;

      return getLatestSFN({ accessKey, secretKey, region, stateMachineArn });
    },

    markStale() {
      self.objects.forEach((object) => {
        object.setStale(true);
      });
    },

    removeStale() {
      self.objects.forEach((object) => {
        if (object.stale) self.objects.delete(object.name); // It is safe to delete from a map while iterating
      });
    },
  }))

  .actions((self) => {
    const superAfterCreate = self.afterCreate;
    const superReset = self.reset;
    return {
      afterCreate() {
        superAfterCreate();
        self._errorDetail = "";
        self.loadingStage = undefined;
      },

      reset() {
        superReset();
        self._errorDetail = "";
        self.objects.clear();
        self.loadingStage = undefined;
      },

      async doLoad({ executionArn, startTime }: { executionArn?: string; startTime?: number } = {}) {
        self.loadingStage = LoadingStage.GettingSchemas;
        self._errorDetail = "";
        self.startTime = startTime ? startTime : self.startTime;
        self.markStale();

        if (!executionArn) {
          const latestExecution = await self.getLatestExecution();
          if (!latestExecution || !latestExecution.executionArn) {
            // If we don't have any, then there is nothing else to do
            self.setStartTime(0);
            self.clearLoadingStage();
            return;
          }
          self.setStartTime(latestExecution.startDate?.getTime() ?? 0);
          executionArn = latestExecution.executionArn;
        }

        // Load the schemas
        await self.loadSchemas();

        let status: ExecutionStatus;
        const shouldStop = (status: ExecutionStatus) => isFailureStatus(status) || isSuccessStatus(status);
        self.setLoadingStage(LoadingStage.MonitoringProgress);

        do {
          status = await self.describeWorkflow(executionArn);
          await delay(10);
          await self.pullProgress(executionArn);
        } while (!shouldStop(status));

        self.clearLoadingStage();
        if (isSuccessStatus(status)) {
          self.removeStale();
          return; // We are done with the success scenario
        }

        // We hit an SFN failure
        await self.handleSFNFailure(executionArn, status);
      },
    };
  });

// see https://mobx-state-tree.js.org/tips/typescript
export interface IImportStatusStore extends Instance<typeof ImportStatusStore> {}

export function useImportStatusStore() {
  const installation = useInstallation();
  const store = installation.importStatusStore;

  useEffect(() => {
    // TODO - add a check for time and if it has been more than 10 minutes, then load the store
    if (!isStoreNew(store)) return;
    store.load();
  }, [store]);

  return { isError: isStoreError(store), store };
}
