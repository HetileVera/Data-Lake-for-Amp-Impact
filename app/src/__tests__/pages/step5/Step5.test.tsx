import { cleanup, screen } from "@testing-library/react";
import { create } from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

import { render, RenderWithChakra } from "test-utils";

import * as appContext from "AppContext";
import * as metadataStore from "models/MetadataStore";
import * as importStatusStore from "models/ImportStatusStore";
import { Step5 } from "pages/step5/Step5";

const spyScrollTo = jest.fn();

describe("Step5 component", () => {
  beforeEach(() => {
    Object.defineProperty(global.window, "scrollTo", { value: spyScrollTo });
  });

  afterEach(() => {
    spyScrollTo.mockClear();
    cleanup;
  });

  test("snapshot", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        id: "1",
        startDate: "2022-01-01",
        startedBy: "test-user",
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        credentials: [],
        nextStepNumber: "2",
        deploymentOperations: {
          isInProgress: true,
          isSuccess: true,
          operations: [],
        },
        deploymentStep: {
          markStarted: jest.fn(),
          markCompleted: jest.fn(),
        },
        triggerDeployment: jest.fn(),
      };
    });
    // @ts-ignore
    jest.spyOn(metadataStore, "useMetadataStore").mockImplementation(() => {
      return {
        isReady: true,
        store: {
          selectedObjects: [{}],
        },
      };
    });
    // @ts-ignore
    jest.spyOn(importStatusStore, "useImportStatusStore").mockImplementation(() => {
      return {
        isError: false,
        store: {
          loadingError: false,
          selectedObjects: [{}],
        },
      };
    });

    const tree = create(
      <RenderWithChakra>
        <appContext.StoreProvider>
          <BrowserRouter>
            <Step5 />
          </BrowserRouter>
        </appContext.StoreProvider>
      </RenderWithChakra>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("render properly", () => {
    // @ts-ignore
    jest.spyOn(appContext, "useInstallation").mockImplementation(() => {
      return {
        id: "1",
        startDate: "2022-01-01",
        startedBy: "test-user",
        accountId: "test-account-id",
        region: "us-east-1",
        appFlowConnectionName: "test-appflow-connection-name",
        credentials: [],
        nextStepNumber: "2",
        deploymentOperations: {
          isInProgress: true,
          isSuccess: true,
          operations: [],
        },
        deploymentStep: {
          markStarted: jest.fn(),
          markCompleted: jest.fn(),
        },
        triggerDeployment: jest.fn(),
      };
    });
    // @ts-ignore
    jest.spyOn(metadataStore, "useMetadataStore").mockImplementation(() => {
      return {
        isReady: true,
        store: {
          selectedObjects: [{}],
        },
      };
    });
    // @ts-ignore
    jest.spyOn(importStatusStore, "useImportStatusStore").mockImplementation(() => {
      return {
        isError: false,
        store: {
          loadingError: false,
          selectedObjects: [{}],
        },
      };
    });

    render(
      <appContext.StoreProvider>
        <BrowserRouter>
          <Step5 />
        </BrowserRouter>
      </appContext.StoreProvider>
    );

    expect(screen.getAllByText(/sit back and relax/i)).toBeTruthy();
    expect(screen.getAllByText(/installation id/i)).toBeTruthy();
    expect(screen.getByText(/we are provisioning the data lake/i)).toBeInTheDocument();
    expect(screen.queryByText("Previous")).not.toBeInTheDocument();
    expect(screen.getByText("Next")).toBeInTheDocument();
  });
});
