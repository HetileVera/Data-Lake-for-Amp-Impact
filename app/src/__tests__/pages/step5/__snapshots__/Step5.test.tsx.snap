// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step5 component snapshot 1`] = `
Array [
  <div
    className="css-0"
    id="step5"
  >
    <div
      className="css-1v55i00"
      data-testid="curved-box"
    />
    <div
      className="chakra-container css-1szj4qk"
    >
      <div
        className="chakra-stack css-olw0lh"
      >
        <div
          className="chakra-stack css-18lyn6q"
        >
          <div
            className="css-1i3fafl"
          >
            <img
              className="chakra-image css-0"
              src="Salesforce_Corporate_Logo_RGB.png"
            />
          </div>
          <div
            className="chakra-stack__divider css-r2vkku"
          />
          <div
            className="css-nutmff"
          >
            <img
              className="chakra-image css-ir5tdx"
              src="AWS_logo_RGB_BLK.png"
            />
          </div>
        </div>
        <div
          className="chakra-stack css-1n7gsjr"
        >
          <div
            className="css-18n4g2t"
          >
            <h2
              className="chakra-heading css-1opsxio"
              id="h2-title"
            >
              Data Lake for Nonprofits, Powered by AWS
            </h2>
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-79elbk"
    >
      <div
        className="css-t7j6xp"
        data-testid="curved-box"
      />
      <div
        className="css-15dboc9"
      >
        <div
          className="chakra-container css-h66qep"
        >
          <div
            className="chakra-stack css-84zodg"
          >
            <span
              className="css-9xsoz7"
            >
              1
            </span>
            <span
              className="css-9xsoz7"
            >
              2
            </span>
            <span
              className="css-9xsoz7"
            >
              3
            </span>
            <span
              className="css-9xsoz7"
            >
              4
            </span>
            <span
              className="css-hfqsm"
            >
              5
            </span>
            <span
              className="css-ulmpab"
            >
              6
            </span>
          </div>
        </div>
      </div>
      <div
        className="chakra-container css-8tuq56"
      >
        <div
          className="css-141wk90"
        >
          <p
            className="chakra-text css-0"
          >
            Installation ID
          </p>
          <p
            className="chakra-text css-0"
          >
            1
          </p>
        </div>
        <h2
          className="chakra-heading css-1vtzlyn"
        >
          Sit back and relax
        </h2>
        <div
          className="css-1fynaiw"
        >
          <div
            className="chakra-stack css-1wy5s38"
          >
            <div
              className="css-12595yo"
            >
              We are provisioning the data lake and importing the data. You can sit back and relax and watch the progress of the data import.
            </div>
            <div
              className="css-12595yo"
            >
              Don't close the browser tab, otherwise the provisioning will pause before it's fully completed.
            </div>
          </div>
        </div>
        <div
          className="css-ib9iq6"
        />
        <div
          className="css-1eqgnj9"
        >
          <div
            className="css-1wzwgz0"
          />
          <div
            className="css-tr1ws2"
          >
            <div
              aria-valuemax={100}
              aria-valuemin={0}
              className="css-z89zyo"
              data-indeterminate=""
              role="progressbar"
              style={
                Object {
                  "width": "0%",
                }
              }
            />
          </div>
          <img
            className="chakra-image css-ju4org"
            src="processing-sforg-01.png"
          />
          <img
            className="chakra-image css-1i9i4ra"
            src="processing-aws-01.png"
          />
          <div
            className="css-z7ff2d"
          />
          <div
            className="css-11aidz8"
          >
            <div
              className="chakra-alert css-b003kq"
              role="alert"
            >
              <div
                className="chakra-alert__desc css-zzks76"
              >
                Don't close the browser tab, otherwise the provisioning will pause before it's fully completed.
              </div>
            </div>
          </div>
          <div
            className="chakra-alert css-1liun4e"
            role="alert"
          >
            <span
              className="chakra-alert__icon css-roekqd"
            >
              <svg
                className="chakra-icon css-fi0ww0"
                focusable={false}
                viewBox="0 0 24 24"
              >
                <path
                  d="M12,0A12,12,0,1,0,24,12,12.014,12.014,0,0,0,12,0Zm6.927,8.2-6.845,9.289a1.011,1.011,0,0,1-1.43.188L5.764,13.769a1,1,0,1,1,1.25-1.562l4.076,3.261,6.227-8.451A1,1,0,1,1,18.927,8.2Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <div
              className="chakra-alert__title css-1vkxl5k"
            >
              Data Lake is ready!
            </div>
            <div
              className="chakra-alert__desc css-1oa6yzk"
            >
              Great news! we got everything in place and you are ready to proceed to the next step
            </div>
          </div>
          <div
            className="css-1pjjc6w"
          >
            <div
              className="css-1o3oomb"
            >
              <div
                className="chakra-stack css-103t9v7"
              >
                <div
                  className="css-1ki54i"
                >
                  <p
                    className="chakra-text css-xytqcl"
                  >
                    NaN
                    %
                  </p>
                  <p
                    className="chakra-text css-0"
                  >
                    Overall Progress
                  </p>
                  <div
                    className="css-2cscy"
                  >
                    <div
                      aria-valuemax={100}
                      aria-valuemin={0}
                      aria-valuenow={NaN}
                      className="css-pez42p"
                      role="progressbar"
                      style={
                        Object {
                          "width": "NaN%",
                        }
                      }
                    />
                  </div>
                </div>
                <div
                  className="css-1budavu"
                >
                  <svg
                    className="chakra-icon css-xcmdz3"
                    focusable={false}
                    viewBox="0 0 24 24"
                  >
                    <g
                      fill="currentColor"
                    >
                      <path
                        d="M12,0A12,12,0,1,0,24,12,12.014,12.014,0,0,0,12,0Zm0,22A10,10,0,1,1,22,12,10.011,10.011,0,0,1,12,22Z"
                      />
                      <path
                        d="M17.134,15.81,12.5,11.561V6.5a1,1,0,0,0-2,0V12a1,1,0,0,0,.324.738l4.959,4.545a1.01,1.01,0,0,0,1.413-.061A1,1,0,0,0,17.134,15.81Z"
                      />
                    </g>
                  </svg>
                  0:00:00
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="css-14wb451"
        >
          <div
            className="chakra-stack css-1y0z9d0"
          >
            <h2
              className="chakra-heading css-lx7khp"
            >
              Latest Data Import
              <span
                className="css-1jdvr3r"
              />
            </h2>
            <div
              className="chakra-stack css-84zodg"
            >
              <p
                className="chakra-text css-bxgqra"
              />
              <button
                aria-label="Check again"
                className="chakra-button css-xroa0q"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="chakra-icon css-onkibi"
                  focusable={false}
                  viewBox="0 0 24 24"
                >
                  <g
                    fill="currentColor"
                  >
                    <path
                      d="M10.319,4.936a7.239,7.239,0,0,1,7.1,2.252,1.25,1.25,0,1,0,1.872-1.657A9.737,9.737,0,0,0,9.743,2.5,10.269,10.269,0,0,0,2.378,9.61a.249.249,0,0,1-.271.178l-1.033-.13A.491.491,0,0,0,.6,9.877a.5.5,0,0,0-.019.526l2.476,4.342a.5.5,0,0,0,.373.248.43.43,0,0,0,.062,0,.5.5,0,0,0,.359-.152l3.477-3.593a.5.5,0,0,0-.3-.844L5.15,10.172a.25.25,0,0,1-.2-.333A7.7,7.7,0,0,1,10.319,4.936Z"
                    />
                    <path
                      d="M23.406,14.1a.5.5,0,0,0,.015-.526l-2.5-4.329A.5.5,0,0,0,20.546,9a.489.489,0,0,0-.421.151l-3.456,3.614a.5.5,0,0,0,.3.842l1.848.221a.249.249,0,0,1,.183.117.253.253,0,0,1,.023.216,7.688,7.688,0,0,1-5.369,4.9,7.243,7.243,0,0,1-7.1-2.253,1.25,1.25,0,1,0-1.872,1.656,9.74,9.74,0,0,0,9.549,3.03,10.261,10.261,0,0,0,7.369-7.12.251.251,0,0,1,.27-.179l1.058.127a.422.422,0,0,0,.06,0A.5.5,0,0,0,23.406,14.1Z"
                    />
                  </g>
                </svg>
              </button>
            </div>
          </div>
          <div
            className="chakra-stat__group css-1t3qzs9"
            role="group"
          >
            <div
              className="chakra-stat css-1wqpdlj"
            >
              <dl>
                <dt
                  className="chakra-stat__label css-1bsgmhw"
                >
                  Objects
                </dt>
                <dd
                  className="chakra-stat__number css-1t2gott"
                >
                  N/A
                </dd>
              </dl>
            </div>
            <div
              className="chakra-stat css-1mbo1ls"
            >
              <dl>
                <dt
                  className="chakra-stat__label css-1bsgmhw"
                >
                  Records
                </dt>
                <dd
                  className="chakra-stat__number css-1t2gott"
                />
              </dl>
            </div>
            <div
              className="chakra-stat css-1mbo1ls"
            >
              <dl>
                <dt
                  className="chakra-stat__label css-1bsgmhw"
                >
                  Fields
                </dt>
                <dd
                  className="chakra-stat__number css-1t2gott"
                />
              </dl>
            </div>
            <div
              className="chakra-stat css-1mbo1ls"
            >
              <dl>
                <dt
                  className="chakra-stat__label css-1bsgmhw"
                >
                  Size
                </dt>
                <dd
                  className="chakra-stat__number css-1t2gott"
                />
              </dl>
            </div>
            <div
              className="chakra-stat css-1mbo1ls"
            >
              <dl>
                <dt
                  className="chakra-stat__label css-1bsgmhw"
                >
                  Time
                </dt>
                <dd
                  className="chakra-stat__number css-1t2gott"
                />
              </dl>
            </div>
          </div>
          <div
            className="css-0"
          >
            <div
              className="css-qwmnd6"
              onClick={[Function]}
            >
              <svg
                className="chakra-icon css-m6c8f6"
                focusable={false}
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  fill="currentColor"
                />
              </svg>
              see more
            </div>
          </div>
        </div>
        <div
          className="chakra-stack css-10428nh"
        >
          <button
            className="chakra-button css-zb27lh"
            disabled={false}
            id="step5-btn-next"
            onClick={[Function]}
            type="button"
          >
            Next
            <span
              className="chakra-button__icon css-1hzyiq5"
            >
              <svg
                aria-hidden={true}
                className="chakra-icon css-onkibi"
                focusable={false}
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
    <span />
  </div>,
  <span />,
]
`;
